sites/cnfpt/src/backend/base/tests/test_cache.py

```diff
@@ -0,0 +1,159 @@
+                "OPTIONS": {
+                    "CLIENT_CLASS": "richie.apps.core.cache.SentinelClient",
+                },
+            },
+            "fallback": {
+                "BACKEND": "django.core.cache.backends.dummy.DummyCache",
+            },
+        }
+    )
+    def test_client(self):
+        """Test class instance of caches"""
+        client = RedisCacheWithFallback(None, {})
+        self.assertIs(type(client), RedisCacheWithFallback)
+        self.assertIs(type(client._redis_cache), RedisCache)
+        self.assertIs(type(client._fallback_cache), DummyCache)
+
+    @mock.patch.object(RedisCacheWithFallback, "_call_fallback_cache")
+    @mock.patch.object(RedisCacheWithFallback, "_call_redis_cache")
+    def test_get_redis_cache(self, redis_cache_mock, fallback_cache_mock):
+        """Test that redis_cache is used by default."""
+        client = RedisCacheWithFallback(None, {})
+        client.get("irrelevent")
+
+        redis_cache_mock.assert_called_once()
+        fallback_cache_mock.assert_not_called()
+
+    @mock.patch("base.cache.logger")
+    @mock.patch.object(RedisCacheWithFallback, "_call_fallback_cache")
+    @mock.patch.object(RedisCacheWithFallback, "_call_redis_cache")
+    def test_get_fallback_cache(
+        self, redis_cache_mock, fallback_cache_mock, logger_mock
+    ):
+        """
+        Test case when redis_cache raises an exception,
+        logger logs the exception then fallback_cache takes over
+        """
+        client = RedisCacheWithFallback(None, {})
+        client.get("irrelevent")
+
+        redis_cache_mock.assert_called_once()
+        fallback_cache_mock.assert_not_called()
+
+        redis_cache_mock.side_effect = Exception()
+        client.get("irrelevent")
+
+        logger_mock.warning.assert_called_with(
+            "[DEGRADED CACHE MODE] - Switch to fallback cache"
+        )
+        logger_mock.exception.assert_called_once()
+        fallback_cache_mock.assert_called_once()
+
+    @override_settings(
+        CACHES={
+            "default": {
+                "BACKEND": "base.cache.RedisCacheWithFallback",
+                "LOCATION": "mymaster/redis-sentinel:26379,redis-sentinel:26379/0",
+                "OPTIONS": {
+                    "CLIENT_CLASS": "richie.apps.core.cache.SentinelClient",
+                },
+            },
+            "fallback": {
+                "BACKEND": "django.core.cache.backends.dummy.DummyCache",
+            },
+        }
+    )
+    @mock.patch.object(DummyCache, "delete")
+    @mock.patch.object(RedisCacheWithFallback, "_call_redis_cache")
+    def test_invalidate_fallback_cache(self, redis_cache_mock, delete_mock):
+        """
+        Test that fallback_cache is invalidated every 60 seconds
+        when Redis is up
+        """
+        now = datetime.datetime.now()
+
+        client = RedisCacheWithFallback(None, {})
+        client.get("round_0")
+        redis_cache_mock.reset_mock()
+        delete_mock.reset_mock()
```

Comment:
As `RedisCacheWithFallback` is the default cache in settings and there are also other tests that do cache access, we cannot check cache is invalidated at the first access cache in this test.